src/Raven.Server/Documents/ShardedHandlers/ShardedBatchHandler.cs

```diff
@@ -0,0 +1,621 @@
+                        }
+                        list.Add((id,expectedChangeVector));
+                    }
+
+                    foreach (var kvp in idsByShard)
+                    {
+                        yield return new SingleShardedCommand
+                        {
+                            Shard = kvp.Key,
+                            CommandStream = bufferedCommand.ModifyBatchPatchStream(kvp.Value),
+                            PositionInResponse = positionInResponse
+                        };
+                    }
+
+                    positionInResponse++;
+                    continue;
+                }
+
+                var shard = _shardedContext.GetShardIndex(_context, cmd.Id);
+                var commandStream = bufferedCommand.CommandStream;
+                var stream = cmd.Type == CommandType.AttachmentPUT ? AttachmentStreams[streamPosition++] : null;
+
+                if (bufferedCommand.IsIdentity)
+                {
+                    commandStream = bufferedCommand.ModifyIdentityStream(cmd.Id);
+                }
+
+                yield return new SingleShardedCommand
+                {
+                    Shard = shard,
+                    AttachmentStream = stream,
+                    CommandStream = commandStream,
+                    PositionInResponse = positionInResponse++
+                };
+            }
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        public void Dispose()
+        {
+        }
+    }
+
+    public class ShardedBatchCommandBuilder : BatchRequestParser.BatchCommandBuilder
+    {
+        public List<Stream> Streams;
+        public List<BufferedCommand> BufferedCommands = new List<BufferedCommand>();
+
+        private readonly bool _encrypted;
+        private readonly ShardedContext _shardedContext;
+
+        public ShardedBatchCommandBuilder(ShardedRequestHandler handler) :
+            base(handler, handler.ShardedContext.DatabaseName, handler.ShardedContext.IdentitySeparator)
+        {
+            _shardedContext = handler.ShardedContext;
+            _encrypted = handler.ShardedContext.Encrypted;
+        }
+
+        public override async Task SaveStream(JsonOperationContext context, Stream input)
+        {
+            Streams ??= new List<Stream>();
+            var attachment = GetServerTempFile("sharded").StartNewStream();
+            await input.CopyToAsync(attachment, Handler.AbortRequestToken);
+            await attachment.FlushAsync(Handler.AbortRequestToken);
+            Streams.Add(attachment);
+        }
+
+        public StreamsTempFile GetServerTempFile(string prefix)
+        {
+            var name = $"{_shardedContext.DatabaseName}.attachment.{Guid.NewGuid():N}.{prefix}";
+            var tempPath = ServerStore._env.Options.DataPager.Options.TempPath.Combine(name);
+            
+            return new StreamsTempFile(tempPath.FullPath, _encrypted);
```

Comment:
Let's create an issue to worry about encrypted sharded databases here, we need to make sure that we'll handle this properly and not write plain text data here. Would also give us a chance to have a temp directory for the sharded db. 

Another thought, do we have a way to control the _paths_ of the sharded dbs?

Reply:
Right now the shards inherit the path from database record.

Opened an issue about encryption https://issues.hibernatingrhinos.com/issue/RavenDB-16277
Not sure what should we consider here.